joml-geometry/src/test/java/org/terasology/joml/geom/RectangleiTest.java

```diff
@@ -2,97 +2,168 @@
-
-        assertTrue(rect.isValid());
-        assertTrue(rect.containsPoint(new Vector2i(0, 0)));
-        assertTrue(rect.containsPoint(new Vector2i(1, 1)));
-        assertFalse(rect.containsPoint(new Vector2i(-1, -1)));
-        assertFalse(rect.containsPoint(new Vector2i(4, 4)));
+class RectangleiTest {
+    static final Vector2ic ZERO = new Vector2i(0, 0);
+    static final Vector2ic MIN = new Vector2i(-1, -1);
+    static final Vector2ic MAX = new Vector2i(1, 1);
+    static final Rectangleic BASE_RECT = new Rectanglei(MIN, MAX);
+
+    static Stream<Arguments> containedRectangles() {
+        return Stream.of(
+                Arguments.of(new Rectanglei(BASE_RECT)),
+                Arguments.of(new Rectanglei(MIN, ZERO)),
+                Arguments.of(new Rectanglei(ZERO, MAX))
+        );
     }
-    @Test
-    public void testRectangleIntersection() {
-        Rectanglei first = new Rectanglei(0, 0, 3, 3);
-        Rectanglei second = new Rectanglei(-1, -1, 2, 2);
-
-        // is valid
-        assertTrue(first.isValid());
-        assertTrue(second.isValid());
 
-        assertFalse(first.containsRectangle(second));
-        assertFalse(second.containsRectangle(first));
+    /**
+     * Rectangles overlapping with BASE_RECT and the intersecting area.
+     */
+    static Stream<Arguments> overlappingRectangles() {
+        return Stream.of(
+                Arguments.of(
+                        new Rectanglei(MIN.x() - 1, MIN.y() - 1, 0, 0),
+                        new Rectanglei(MIN, ZERO)),
+                Arguments.of(
+                        new Rectanglei(0, 0, MAX.x() + 1, MAX.y() + 1),
+                        new Rectanglei(ZERO, MAX))
+        );
+    }
 
-        assertTrue(first.intersectsRectangle(second));
-        assertTrue(second.intersectsRectangle(first));
-        assertEquals(first.intersection(second, new Rectanglei()), new Rectanglei(0, 0, 2, 2));
+    /**
+     * Rectangles touching one side of BASE_RECT.
+     */
+    static Stream<Arguments> touchingRectangles() {
+        Rectanglei right = new Rectanglei(MAX.x(), MIN.y(), MAX.x() + 1, MAX.y());
+        Rectanglei left = new Rectanglei(MIN.x() - 1, MIN.y(), MIN.x(), MAX.y());
+        Rectanglei top = new Rectanglei(MIN.x(), MAX.y(), MAX.x(), MAX.y() + 1);
+        Rectanglei bottom = new Rectanglei(MIN.x(), MIN.y() - 1, MAX.x(), MIN.y());
+        return Stream.of(
+                Arguments.of(right),
+                Arguments.of(left),
+                Arguments.of(top),
+                Arguments.of(bottom)
+        );
+    }
 
+    static Stream<Arguments> invalidRectangles() {
+        return Stream.of(
+                Arguments.of(new Rectanglei()),                                     // invalid by default
+                Arguments.of(new Rectanglei(ZERO)),                                 // a "point"
+                Arguments.of(new Rectanglei(MIN.x(), MIN.y(), MAX.x(), MIN.y()))    // a "line"
```

Comment:
>  I also think that a rectangle with 0 area should be valid too

This would mean that there are no invalid rectangles, and we could just remove that property altogether, wouldn't it?

Reply:
No, a rectangle where either of the lengths is strictly less than 0 would still be invalid.